### PR TITLE
Update PSM To Have Separate Mint/Redeem Pausing

### PR DESCRIPTION
**Pull request by ElliotFriedman**
use 1 oracle as fixed price oracle in OracleRef and the actualPriceOracle in the SafePSM to check whether or not the other oracle is within the PriceBoundPSM safe bounds.

### CODE REVIEW

File: contracts/peg/SafePSM.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        require(valid, "SafePSM: oracle price invalid");
+
+        return _peg;
+    }
+
+    /// @notice function that reverts if the oracle price is outside of
+    /// the allowable price range
+    function _validatePrice() internal view {
+        _validatePriceRange(readActualOracle());
+    }
```
Comment (Joeysantoro) [inline contracts/peg/SafePSM.sol:51]: wouldn't it make sense for this to override the existing logic rather than be in a new hook?

Comment (ElliotFriedman) [reply contracts/peg/SafePSM.sol]: moved things around, the issue is that the existing hook gets called with the constant oracle value

File: contracts/peg/SafePSM.sol [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        _underlyingToken,
+        _surplusTarget
+    ) {
+        actualPriceOracle = _actualPriceOracle;
+    }
+
+    /// @notice the price of the referenced oracle
+    /// @return the peg as a Decimal
+    /// @dev the peg is defined as FEI per X with X being ETH, dollars, etc
+    function readActualOracle() public view returns (Decimal.D256 memory) {
```
Comment (eswak) [inline contracts/peg/SafePSM.sol:40]: I think you should use the normal oracle & instead override _getMintAmountOut and _getRedeemAmountOut to not read the oracle value (that will be fixed to 1 anyway)
you don't need the oracle that returns 1
and then you can keep the default behavior on _validatePriceRange & _validPrice
```
function _getMintAmountOut(uint256 amountIn) private view returns (uint256 amountFeiOut) {
        amountFeiOut = amountIn
            .mul(Constants.BASIS_POINTS_GRANULARITY - mintFeeBasisPoints)
            .div(Constants.BASIS_POINTS_GRANULARITY)
            .asUint256();
    }
```
something like that

File: contracts/peg/PegStabilityModule.sol [hunk 3]
```diff
@@ -68,6 +86,42 @@ contract PegStabilityModule is IPegStabilityModule, RateLimitedMinter, OracleRef
...
     }
 
+    /// @notice modifier that allows execution when redemptions are not paused
+    modifier whileRedemptionsNotPaused {
+        require(!redeemPaused && !paused(), "PegStabilityModule: Redeem paused");
+        _;
+    }
+
+    /// @notice modifier that allows execution when minting is not paused
+    modifier whileMintingNotPaused {
```
Comment (ElliotFriedman) [inline contracts/peg/PegStabilityModule.sol:96]: split out modifiers

File: contracts/peg/PegStabilityModule.sol [hunk 4]
```diff
@@ -68,6 +86,42 @@ contract PegStabilityModule is IPegStabilityModule, RateLimitedMinter, OracleRef
...
+        require(!mintPaused && !paused(), "PegStabilityModule: Minting paused");
+        _;
+    }
+
+    /// @notice set secondary pausable methods to paused
+    function pauseRedeem() public isGovernorOrGuardianOrAdmin {
```
Comment (ditchfieldcaleb) [inline contracts/peg/PegStabilityModule.sol:102]: make external

Comment (ditchfieldcaleb) [reply contracts/peg/PegStabilityModule.sol:102]: + others external too

**Review by eswak — DISMISSED** [review]

**Review by Joeysantoro — DISMISSED** [review]

File: contracts/peg/SafePSM.sol (hunk 2 above)
Comment (ElliotFriedman) [reply contracts/peg/SafePSM.sol]: will address this in a separate pr

File: proposals/dao/fip_72.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+Deploy Steps:
+  0 - Deploy New FixedPricePSM
+*/
+
+const fipNumber = '72';
+
+// mint fee starts at 0
+const daiPSMMintFeeBasisPoints = 0;
+// redeem fee is 30 basis points
```
Comment (Joeysantoro) [inline proposals/dao/fip_72.ts:25]: comment off

File: proposals/description/fip_72.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      target: 'core',
+      values: '0',
+      method: 'grantMinter(address)',
+      arguments: ['{daiFixedPricePSM}'],
+      description: 'Grant DAI psm minter role'
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'revokeMinter(address)',
```
Comment (Joeysantoro) [inline proposals/description/fip_72.ts:16]: remove for rollback potential

File: proposals/description/fip_72.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    }
+  ],
+  description: `
+Forum discussion: https://tribe.fei.money/t/fip-72-peg-fei-to-dai/3905/11
+Snapshot: https://snapshot.fei.money/#/proposal/0xbb8160b0835556d2472c27bd05750bb3060a2f4377bcb808d10daaeb603b117b
+
+- Grant new DAI PSM minter role
+- Revoke minter role from old DAI psm
+- Set target of DAI PCV Drip Controller to new DAI PSM
+- Grant new DAI PSM FEI skimmer PCV controller role
```
Comment (Joeysantoro) [inline proposals/description/fip_72.ts:84]: there is no new

File: proposals/description/fip_72.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+- Remove old DAI PSM from pcv guardian
+- Remove all FEI and DAI from old DAI PSM and deposit it into new DAI PSM
+- Add new DAI PSM to pcv guardian as a safe address
+- Add new DAI PSM to CR Oracle
+- Remove old DAI PSM from CR Oracle
+- Pause DAI PSM
```
Comment (Joeysantoro) [inline proposals/description/fip_72.ts:90]: Deprecate old is clearer

**Review by Joeysantoro — APPROVED** [review]

**Review by Joeysantoro — APPROVED** [review]

**Pull request merged**